newsService/src/main/java/com/kovan/app/service/NewsService.java

```diff
@@ -33,51 +39,93 @@ public class NewsService {
         this.service = service;
         this.newsRepository = newsRepository;
+        this.newsMapper = newsMapper;
     }
-
-    public NewsDto getTopHeadlines() {
-
-        List<NewsDto.Article> articles = new ArrayList<>();
-        int pageSize = 0;
-        int page = 1;
-        NewsDto newsDto;
-
-        do {
-            pageSize += 20;
-            String apiUrl = buildUrl(page++);
-            String response;
-            try {
-                response = restTemplate.getForObject(apiUrl, String.class);
-                newsDto = objectMapper.readValue(response, NewsDto.class);
-                articles.addAll(newsDto.getArticles());
-            } catch (JsonProcessingException e) {
-                throw new NewsRetrievalException("Failed to parse news data from API response.", e);
+    public List<NewsDto> getTopHeadlines() {
+
+        AtomicInteger PAGE_SIZE = new AtomicInteger(0);
+
+        AtomicReference<List<NewsDto.Article>> articles= new AtomicReference<>();
+
+        iterate(1, page -> page + 1)
+                .map(page -> {
+                    String apiUrl = buildUrl(page);
+                    String response;
+                    try {
+                        response = restTemplate.getForObject(apiUrl, String.class);
+                        NewsDto newsDto =  objectMapper.readValue(response, NewsDto.class);
+                        System.out.println(page);
+                        if(isNull(articles.get())) {
+                            articles.set(new ArrayList<>());
+                        }
+                        articles.get().addAll(newsDto.getArticles());
+                        return newsDto;
+                    } catch (JsonProcessingException e) {
+                        throw new NewsRetrievalException("Failed to parse news data from API response.", e);
+                    }
+                })
+                .takeWhile(newsDto -> PAGE_SIZE.addAndGet(DEFAULT_PAGE_SIZE) < newsDto.getTotalResults())
+                .toList();
+
+        boolean isYesterdayInDb = isNewsAlreadyInDb(yesterday.toString());
+        boolean isTodayInDb = isNewsAlreadyInDb(today.toString());
+
+        Map<Boolean, List<NewsDto.Article>> partitionedArticles = articles.get().stream()
+                .takeWhile(article -> !parseDate(article.getPublishedAt()).isBefore(yesterday))
+                .collect(partitioningBy(article -> parseDate(article.getPublishedAt()).equals(yesterday)));
+
+        List<NewsDto.Article> newsArticles = partitionedArticles.get(true);
+        List<NewsDto.Article> futureArticles = partitionedArticles.get(false);
+
+        if (newsArticles.isEmpty() && futureArticles.isEmpty() && !isYesterdayInDb) {
+            NewsDto emptyNews = NewsDto.builder().totalResults(0)
+                    .publishedAt(yesterday.toString()).status("fail").build();
+            service.saveNewsInDb(emptyNews);
+        } else {
+            if (!isYesterdayInDb && !newsArticles.isEmpty()) {
```

Comment:
Avoid code Repetition try to find better solution

Reply:
Refactored the code by creating separate method for handle save news and avoid redundancy.